### PR TITLE
Add support for passthru of sign argument

### PR DESCRIPTION
**Pull request by CleyFaye**
These changes adds an extra parameter to sign() to control the signature behavior. In particular it allows the production of non-deterministic signatures by using randomness.

This does not break compatibility (if the argument is missing previous behavior is maintained).

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -44,9 +44,13 @@ function magicHash (message, messagePrefix) {
   return hash256(buffer)
 }
 
-function sign (message, privateKey, compressed, messagePrefix) {
+function sign (message, privateKey, compressed, messagePrefix, sigOptions) {
+  if ((typeof messagePrefix) === 'object' && sigOptions === undefined) {
```
Comment (fanatid) [inline index.js]: extra parentheses, can be removed

**Comment by fanatid** [comment]
Code looks fine. Tests should be fixed after https://github.com/bitcoinjs/bs58check/pull/20

**Comment by CleyFaye** [comment]
I removed the extra parentheses.

**Comment by fanatid** [comment]
@CleyFaye can you rebase? Test should be fixed with new `bs58check`

**Comment by CleyFaye** [comment]
Sorry, I'm not sure what you mean with rebase. I bumped bs58check to the last version, but travis still fail with old node version (0.x).

I can also update .travis.yml to remove these old version from the tests (and maybe add node 8 too).

**Comment by fanatid** [comment]
Sorry, I was wrong about rebase. You just need recommit for restarting travis. Now related that I can did it on my side by hands :(

**Comment by fanatid** [comment]
Tests fail because old version need safe-buffer in bitcoinjs-message itself...

**Comment by CleyFaye** [comment]
Since node 0.12 is beyond end of life, maybe it can be removed from tests ?

**Comment by fanatid** [comment]
Probably yes, more over, since node 4 already is not supported (end-of-life). Maybe it costs remove `safe-buffer` everywhere? :D ping @dcousens 
I created PR with safe-buffer for this repo in case you if will not want drop old node versions from travis https://github.com/bitcoinjs/bitcoinjs-message/pull/11

@CleyFaye thanks for you efforts, let's wait Daniel's answer

**Comment by dcousens** [comment]
@fanatid happy to remove `0.12` from tests, keeping with the same standards we do elsewhere, LTS up to latest.

Is `safe-buffer` unnecessary now?

**Comment by fanatid** [comment]
Then let's remove old node versions :)

If I remember correctly, functional of safe-buffer was added to node@6.0.0 and bakported to 4 since 4.4.0
So probably yes, if we support only LTS, it should be safe remove this package, right?

**Comment by dcousens** [comment]
@fanatid if we start now,  that is OK,  but lets assess how many packages need to be changed.

**Comment by fanatid** [comment]
Yes... a lot. So, what about drop old node versions here and not add safe-buffer?

**Comment by dcousens** [comment]
Works for me.

**Comment by CleyFaye** [comment]
Should I do it? I already tested by removing node < 4 from travis.yml and adding node 8, I only have to push :)

**Comment by junderw** [comment]
Yeah I went ahead and tried to recreate the object paradigm you had going.

One thing I changed: rather than exposing the entire secp256k1 options arg to the consumer by passing it through, I restricted it only to allow them to pass the data option.

**Comment by CleyFaye** [comment]
This totally address the issue we had, so it's fine.

Thanks ☺

**Pull request merged**